storage/src/journal/fixed.rs

```diff
@@ -0,0 +1,756 @@
+                .truncate(blob_len - blob_len % chunk_size as u64)
+                .await?;
+            newest_blob.sync().await?;
+        }
+
+        Ok(Self {
+            runtime,
+            cfg,
+            blobs,
+            tracked,
+            synced,
+            pruned,
+        })
+    }
+
+    /// Sync any pending updates to disk.
+    pub async fn sync(&mut self) -> Result<(), Error> {
+        self.synced.inc();
+        let newest_blob = self.newest_blob();
+        debug!("syncing blob {}", newest_blob.0);
+        self.newest_blob().1.sync().await.map_err(Error::Runtime)
+    }
+
+    /// Returns the total number of items in the journal, ignoring any pruning. The next value
+    /// appended to the journal will be at this position.
+    pub async fn size(&self) -> Result<u64, Error> {
+        let chunk_size = size_of::<u32>() + N;
+        let newest_blob = self.newest_blob();
+        let blob_len = newest_blob.1.len().await?;
+        assert_eq!(blob_len % chunk_size as u64, 0);
+        let items_in_blob = blob_len / chunk_size as u64;
+        Ok(items_in_blob + self.cfg.items_per_blob * newest_blob.0)
+    }
+
+    /// Append a new item to the journal. Return the item's position in the journal, or error if the
+    /// operation fails.
+    pub async fn append(&mut self, item: [u8; N]) -> Result<u64, Error> {
+        let chunk_size = size_of::<u32>() + N;
+        let mut newest_blob = self.newest_blob();
+
+        let mut blob_len = newest_blob.1.len().await?;
+        assert_eq!(blob_len % chunk_size as u64, 0);
+        let items_in_blob = blob_len / chunk_size as u64;
+
+        // if the blob is full we need to create the next one and use that instead
+        if items_in_blob >= self.cfg.items_per_blob {
+            let next_blob_index = newest_blob.0 + 1;
+            debug!("creating next blob {}", next_blob_index);
+            assert_eq!(items_in_blob, self.cfg.items_per_blob);
+            // always sync the previous blob before creating a new one
+            newest_blob.1.sync().await?;
+            let next_blob = self
+                .runtime
+                .open(&self.cfg.partition, &next_blob_index.to_be_bytes())
+                .await?;
+            assert!(self.blobs.insert(next_blob_index, next_blob).is_none());
+            newest_blob = self.newest_blob();
+            self.tracked.inc();
+            blob_len = 0;
+        }
+
+        let mut buf: Vec<u8> = Vec::with_capacity(chunk_size);
+        let checksum = crc32fast::hash(&item[..]);
+        buf.put(&item[..]);
+        buf.put_u32(checksum);
+
+        let item_position = blob_len / chunk_size as u64;
+        newest_blob.1.write_at(&buf, blob_len).await?;
+        trace!(
+            blob = newest_blob.0,
+            position = item_position,
+            "appended item"
+        );
+        Ok(item_position + self.cfg.items_per_blob * newest_blob.0)
+    }
+
+    /// Read the item at the given position in the journal.
```

Comment:
nit: recommend also clarifying this is the "itemth" position not the byte position.